### PR TITLE
feat: create wallet transactions and invoice

### PR DESCRIPTION
**Pull request by lovrocolic**
## Context

Handle paid credits

## Description

Once we create a wallet or apply top up we need to handle paid credits
- handle paid credits -> generate invoice and add to wallet's balance once payment status is succeeded
- in order to generate invoice we need to create new invoice type
- I have added polymorphic association on `fee` (discussed with Jeremy) so that we don't need to create new association for each new invoice type
- migration of old data will be handled in separate tech task

### CODE REVIEW

File: app/services/invoices/payments/stripe_service.rb [hunk 1]
```diff
@@ -155,6 +157,21 @@ def update_invoice_status(status)
...
+      def update_wallet(status)
+        return unless invoice.invoice_type == 'credit'
+        return unless status == 'succeeded'
+
+        wallet_transaction = invoice.fees.find_by(fee_type: 'credit')&.invoiceable
+
+        return unless wallet_transaction
+        return if wallet_transaction.status == 'settled'
+
+        wallet_transaction.update!(status: :settled, settled_at: Time.zone.now)
```
Comment (jdenquin) [inline app/services/invoices/payments/stripe_service.rb:169]: can we do this a in a separated service? the invoice stripe service should keep his own scope and does not update other objects than invoices

**Review by jdenquin — CHANGES_REQUESTED** [review]
Looks good for me!
This will need a serious QA!, especially around the new fee type.

Just one comment for a change and all good!

File: app/services/invoices/payments/stripe_service.rb (hunk 1 above)
Comment (lovrocolic) [reply app/services/invoices/payments/stripe_service.rb]: Refactored

File: app/services/invoices/paid_credit_service.rb [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    attr_accessor :customer, :date, :wallet_transaction
+
+    def currency
+      @currency ||= customer.default_currency
+    end
+
+    def compute_amounts(invoice)
+      fee_amounts = invoice.fees.select(:amount_cents, :vat_amount_cents)
+
+      invoice.amount_cents = fee_amounts.sum(&:amount_cents)
```
Comment (rsempe) [inline app/services/invoices/paid_credit_service.rb:53]:
```suggestion
      invoice.amount_cents = fee_amounts.sum(:amount_cents)
```

File: app/services/invoices/paid_credit_service.rb [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+      invoice.amount_currency = currency
+      invoice.vat_amount_cents = fee_amounts.sum(&:vat_amount_cents)
```
Comment (rsempe) [inline app/services/invoices/paid_credit_service.rb:55]:
```suggestion
      invoice.vat_amount_cents = fee_amounts.sum(:vat_amount_cents)
```

File: app/models/wallet.rb [hunk 4]
```diff
@@ -21,6 +21,8 @@ def mark_as_terminated!(timestamp = Time.zone.now)
     terminated!
   end
 
+  scope :expired, -> { where('wallets.expiration_date < ?', Time.zone.now.beginning_of_day,) }
```
Comment (rsempe) [inline app/models/wallet.rb:24]:
```suggestion
  scope :expired, -> { where('wallets.expiration_date < ?', Time.current.beginning_of_day,) }
```

File: app/services/invoices/paid_credit_service.rb (hunk 2 above)
Comment (lovrocolic) [reply app/services/invoices/paid_credit_service.rb]: Removed

File: app/services/invoices/paid_credit_service.rb (hunk 3 above)
Comment (lovrocolic) [reply app/services/invoices/paid_credit_service.rb]: Removed

File: app/models/wallet.rb (hunk 4 above)
Comment (lovrocolic) [reply app/models/wallet.rb]: Fixed, actually wasn't aware before`.current` covers it :)

File: app/services/fees/paid_credit_service.rb [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+      currency = invoice.amount.currency
+      rounded_amount = wallet_transaction.amount.round(currency.exponent)
+      amount_cents = rounded_amount * currency.subunit_to_unit
+
+      new_fee = Fee.new(
+        invoice: invoice,
+        fee_type: :credit,
+        invoiceable_type: 'WalletTransaction',
+        invoiceable_id: wallet_transaction.id,
```
Comment (vincent-pochet) [inline app/services/fees/paid_credit_service.rb:23]:
```ruby
        invoiceable: wallet_transaction
```

**Review by jdenquin — APPROVED** [review]
:rocket:

**Review by vincent-pochet — APPROVED** [review]

File: app/services/fees/add_on_service.rb [hunk 6]
```diff
@@ -18,6 +18,9 @@ def create
         applied_add_on: applied_add_on,
         amount_cents: applied_add_on.amount_cents,
         amount_currency: plan.amount_currency,
+        fee_type: :add_on,
+        invoiceable_type: 'AppliedAddOn',
+        invoiceable_id: applied_add_on.id,
```
[the remark this answers is not in the document]
Comment (lovrocolic) [reply app/services/fees/add_on_service.rb]: fixed

File: app/services/fees/charge_service.rb [hunk 7]
```diff
@@ -51,6 +51,9 @@ def init_fee
         amount_cents: amount_cents,
         amount_currency: charge.amount_currency,
         vat_rate: customer.applicable_vat_rate,
+        fee_type: :charge,
+        invoiceable_type: 'Charge',
+        invoiceable_id: charge.id,
```
[the remark this answers is not in the document]
Comment (lovrocolic) [reply app/services/fees/charge_service.rb]: fixed

File: app/services/fees/subscription_service.rb [hunk 8]
```diff
@@ -21,6 +21,9 @@ def create
         amount_cents: new_amount_cents.round,
         amount_currency: plan.amount_currency,
         vat_rate: customer.applicable_vat_rate,
+        fee_type: :subscription,
+        invoiceable_type: 'Subscription',
+        invoiceable_id: subscription.id,
```
[the remark this answers is not in the document]
Comment (lovrocolic) [reply app/services/fees/subscription_service.rb]: fixed

File: app/services/fees/paid_credit_service.rb (hunk 5 above)
Comment (lovrocolic) [reply app/services/fees/paid_credit_service.rb]: fixed

File: app/jobs/bill_paid_credit_job.rb [hunk 9]
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+class BillPaidCreditJob < ApplicationJob
+  queue_as 'billing'
+
+  retry_on Sequenced::SequenceError
+
+  def perform(wallet_transaction, date)
```
[the remark this answers is not in the document]
Comment (lovrocolic) [reply app/jobs/bill_paid_credit_job.rb]: switched to timestamp

**Pull request merged**